File: scripts/test_artifacts.py
"""Output-location policy for generated test artifacts; importing this module has no side effects."""
from datetime import datetime, timezone
from pathlib import Path
from uuid import uuid4


def artifact_directory(project_root, requested=None):
    root = Path(project_root).resolve()
    if requested is None:
        run_id = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ") + "-" + uuid4().hex[:8]
        candidate = root / "test-results" / "tui" / run_id
    else:
        candidate = Path(requested).expanduser().resolve()
    # Resolve both operands, including existing symlinks, before enforcing containment.
    candidate = candidate.resolve()
    allowed = root / "test-results"
    if candidate == root or root in candidate.parents:
        if candidate != allowed and allowed not in candidate.parents:
            raise ValueError("Generated test output inside this repository must be under test-results/, not documentation, fixtures, or source directories.")
    return candidate


def reserve_recording_directory(project_root, requested=None):
    directory = artifact_directory(project_root, requested)
    if directory.exists() and (not directory.is_dir() or any(directory.iterdir())):
        raise ValueError("The recording output directory must be new or empty. Choose a new run directory instead of overwriting evidence.")
    directory.mkdir(parents=True, exist_ok=True)
    try:
        # Reserve the run atomically so concurrent recorders cannot share an empty path.
        with (directory / "run-metadata.json").open("x") as metadata:
            metadata.write("{}\n")
    except FileExistsError as error:
        raise ValueError("This recording directory has already been reserved. Choose a new run directory.") from error
    return directory

File: scripts/test-artifacts.ts
import { existsSync, realpathSync } from "node:fs";
import { dirname, isAbsolute, join, relative, resolve } from "node:path";

function canonical(path: string): string {
  const absolute = resolve(path);
  if (existsSync(absolute)) return realpathSync(absolute);
  const parent = dirname(absolute);
  return parent === absolute ? absolute : join(canonical(parent), relative(parent, absolute));
}
function contained(root: string, path: string): boolean {
  const child = relative(root, path);
  return child === "" || (!child.startsWith(`..${process.platform === "win32" ? "\\" : "/"}`) && child !== ".." && !isAbsolute(child));
}
/** Keep generated outputs out of documentation/source trees, including through symlinks. */
export function artifactDirectory(projectRoot: string, requested: string): string {
  const root = realpathSync(projectRoot);
  const output = canonical(requested);
  if (contained(root, output) && !contained(join(root, "test-results"), output)) {
    throw new Error("Generated test output inside this repository must be under test-results/, not documentation, fixtures, or source directories.");
  }
  return output;
}
